contextualized/regression/datamodules.py

```diff
@@ -0,0 +1,133 @@
+from abc import abstractmethod
+import numpy as np
+import torch
+import pytorch_lightning as pl
+from contextualized.regression.datasets import DATASETS
+
+from torch.utils.data import DataLoader
+
+
+class RegressionDataModule(pl.LightningDataModule):
+    """
+    Torch Datamodule used for contextualized.regression modules
+    """
+
+    def __init__(
+        self,
+        c,
+        x,
+        y,
+        dataset="multivariate",
+        n=0,
+        num_workers=0,
+        batch_size=32,
+        correlation=False,
+        markov=False,
+        **kwargs,
+    ):
+
+        """Initialize the Regression Datamodule
+
+        Args:
+            c (ndarray): 2D array containing contextual features per each sample.
+            x (ndarray): 2D array containing features per each sample.
+            w (ndarray): 3D array containing known 2D network per each sample.
+            dataset (str): Which dataset to use. Choose between ["multivariate", "univariate", "multitask_multivariate", "multitask_univariate].
+            n (int): Number of data samples to use. Defaults to 0 (full dataset will be used).
+            correlation (bool): Whether datamodule will be used for correlation regression module.
+            markov (bool): Whether datamodule will be used for markov regression module. (Currently unused)
+            num_workers (int): Number of workers used in dataloaders.
+            batch_size (int): Size of batches used in dataloaders.           
+        """
+
+        super().__init__()
+
+        self.dataset = DATASETS[dataset]
+        self.num_workers = 0
+        self.batch_size = batch_size
+        if n == 0:
+            n = c.shape[0]  # use full dataset
+
+        self.n = n
+        self.C = torch.tensor(c)
+        self.X = torch.tensor(x)
+        self.Y = torch.tensor(y)
+
+        if correlation or markov:
+            self.Y = self.X
+
+        # partition data
+        self.train_idx, self.test_idx, self.val_idx = self._create_idx(self.n)
+        self.full_dataset = self.dataset(self.C, self.X, self.Y, idx=range(self.n - 80))
+        self.train_dataset = self.dataset(self.C, self.X, self.Y, idx=self.train_idx)
+        self.test_dataset = self.dataset(self.C, self.X, self.Y, idx=self.test_idx)
+        self.val_dataset = self.dataset(self.C, self.X, self.Y, idx=self.val_idx)
+        # self.pred_dataset = self.dataset(self.C, self.X, self.Y, idx=self.pred_idx)
+
+    def setup(self, stage: str):
```

Comment:
In the example code that torch provides, setup is one of the functions that I think they mention is included in order to have operations specific to each of the different runs. Not sure if this would be useful or not in the future for our purposes.

Reply:
I see. Sounds like we should use this to control which dataloader is being referenced for loading samples, which would address some of the other comments.

Reply:
Sounds like a great idea. In reference to your other comment, pytorch does seem to require a predict_dataloader, so I have the “setup” function now being used to swap which of the three datasets is used by predict_dataloader. I’ve so far left out stages “train/test/val” from the setup function, as they aren’t needed yet and to minimize holding onto excess variables. Currently troubleshooting some bugs, but I’ll let you know when I can update the repo to reflect these changes.